lib/api/core/plugins/manifest.js

```diff
@@ -0,0 +1,108 @@
+/*
+ * Kuzzle, a backend software, self-hostable and ready to use
+ * to power modern apps
+ *
+ * Copyright 2015-2018 Kuzzle
+ * mailto: support AT kuzzle.io
+ * website: http://kuzzle.io
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+const
+  _ = require('lodash'),
+  { PluginImplementationError } = require('kuzzle-common-objects').errors,
+  path = require('path'),
+  semver = require('semver');
+
+/* eslint-disable no-console */
+
+/**
+ * This class initializes itself using a manifest.json file
+ *
+ * @deprecated  - the "fallback" argument is deprecated: there will be no
+ * fallback if a manifest.json file is missing in the next major release of
+ * Kuzzle. For now, the fallback is a function receiving this instance as
+ * an argument, and is in charge of updating it accordingly
+ *
+ * @param {Kuzzle} kuzzle       - Kuzzle instance
+ * @param {string} pluginPath   - Absolute path to the plugin directory
+ * @param {Function} [fallback] - Fallback function, if no manifest.json is found
+ */
+class Manifest {
+  constructor(kuzzle, pluginPath, fallback = null) {
+    this.path = pluginPath;
+    this.name = null;
+    this.privileged = false;
+
+    // Only Kuzzle 1.x should support plugins without a kuzzleVersion property set
+    this.kuzzleVersion = '>=1.0.0 <2.0.0';
+
+    let raw;
+    try {
+      raw = require(path.resolve(this.path, 'manifest.json'));
+    } catch (e) {
+      // @deprecated
+      if (fallback !== null) {
+        console.warn(`[${this.path}] Plugins without a manifest.json file are deprecated.`);
+        fallback(this);
+        return this;
+      }
+
+      throw new PluginImplementationError(e);
+    }
+
+    if (!_.isNil(raw.privileged)) {
```

Comment:
Protocols are per definition privileged (and are not plugins).
If we really want to go with a manifest class, I'd use a separate implementation, possibly implementing a common abstract class for plugins vs protocols.

Reply:
* Disagree: why are protocols "per definition" privileged? It seems to me that no one should need to access the `kuzzle` internal object. In fact, I believe in that so strongly that I want to propose to remove "privileged mode" altogether from Kuzzle from the next major release. If more accessors are needed because we didn't anticipate some use case, then we shall add these specifically
* Agree: protocols are not plugins. I will add an abstract class for manifests. This will resolve at least one of your comments below.

Reply:
I don't know any library that tries to hide its internal to developers as strong as we do.
My position is:
1. Why would we block developers as long as it is clear undocumented features are not supported?
2. Javascript is a nice language that in most cases allows developers to achieve what they want eventually
3. Why spend some time & effort in "protecting" our objects?

About the "why": protocols are given an entrypoint instance. Entrypoints rely on kuzzle instance. So protocols can access it (but this is not supported).
The concept of "privileged" is actually meaningless for protocols.

Reply:
@benoitvidis Any library in OOP languages will have protected/private properties, allowing to expose a clean and simple API to its users, while keeping implementation details "hidden" to those who explicitly inherit those objects. Usually, unless their purpose is to be extended, these implementation details are poorly documented, oftentimes on purpose: you do not want to provide support to anyone who want to extend an ever-evolving object.

And we come to the crux of the problem you state: there is no such mechanism in Javascript, everything is public. This is neither good or bad, it's simply different. But there is the problem of how to define a clear line, that is easy to understand for our users, between what is stable for an entire major version, and what one should avoid to use, unless at his own risks.

This is why we provide dedicated, stable and well-documented interfaces for plugins and protocols developers. 
About those, our (my :disappointed: ) first approach was to protect everything, and we clearly overdid it. We stepped back a bit from that position, the only well-protected objects remaining are the Request ones, because securing them allow us to spare a lot of redundant tests to make sure those are correct. And we did that only because you recently found a way to secure them without diminshing their usability: as discussed together, I was ready to un-protect them, since as you put it, in Javascript it's the users responsibility to not break what a library provides.

Do we have to over-protect everything? It's been a while that I think that we went over the top, so my position is a lot more lax now. But do we need to expose internal objects without purpose? I think not.

So, where does that leave us with privileged mode?
As I said, I want to remove this entirely, I see no purpose in int.
And if someone finds a `kuzzle` object by inspecting the innerworks of the interfaces we provide, then so be it. It's their resposibility, and we should not spend time going againt Javascript itself.